### PR TITLE
feat(summarize): Gather failed metrics instead of raising

### LINKED ISSUE #3085: allow catching exceptions raised by metrics
when a metric fail the error is propagated. this prevents seeing the other metric values, or even printing the report object as its repr computes the metrics. this can be annoying, especially when the failing metric is one of the built-in ones that the user did not ask for or is not expecting to work on their data

it would be nice to have an option to instead show "failed" in the output table where the value should go, but not propagate the error. it would still be possible to turn this off so that it is possible to see the full traceback and debug, or to handle the exception however the user sees fit

### PR DESCRIPTION
**Pull request by auguste-probabl**
Closes #3085 

Adds a new attribute `errors` to `MetricSummaryDisplay` which lists all the metric errors so they can be displayed in the `repr`.
Failed metrics remain in the output DataFrame with their value set to `NaN`.

<details>

<summary>Manual tests to showcase the new behavior</summary>

`EstimatorReport`
```py
from sklearn.ensemble import RandomForestClassifier

import skore
import sklearn

X, y = sklearn.datasets.make_classification(n_samples=1000, return_X_y=True)

model = RandomForestClassifier()
report = skore.evaluate(model, X, y)

def fail(e,x,y):
    raise Exception("hey")

report.metrics.add(fail)

report.metrics.summarize()
#                         RandomForestClassifier
# Metric           Label
# Fail                                       NaN
# Accuracy                              0.945000
# Precision        0                    0.953704
#                  1                    0.934783
# Recall           0                    0.944954
#                  1                    0.945055
# ROC AUC                               0.987801
# Log loss                              0.174793
# Brier score                           0.048785
# Fit time (s)                          0.159072
# Predict time (s)                      0.004261
# Use .frame() to control the format of the output.
```

---

`EstimatorReport`, computing metrics on train set with no train set
```py
from sklearn.ensemble import RandomForestClassifier

import skore
import sklearn

X, y = sklearn.datasets.make_classification(n_samples=1000, return_X_y=True)

model = RandomForestClassifier().fit(X, y)
report = skore.EstimatorReport(model, X_test=X, y_test=y)

def fail(e,x,y):
   raise Exception("hey")

report.metrics.add(fail)

report.metrics.summarize(data_source="train")
# Empty DataFrame
# Columns: []
# Index: [Fail, Accuracy, Precision, Recall, ROC AUC, Log loss, Brier score, Fit time (s), Predict time (s)]
# Use .frame() to control the format of the output.
# Metric 'fail' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'accuracy' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'precision' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'recall' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'roc_auc' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'log_loss' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'brier_score' has failed: ValueError('No train data were provided when creating the report.')

report.metrics.summarize(data_source="both")
#                         RandomForestClassifier (test)
# Metric           Label
# Accuracy                                     1.000000
# Precision        0                           1.000000
#                  1                           1.000000
# Recall           0                           1.000000
#                  1                           1.000000
# ROC AUC                                      1.000000
# Log loss                                     0.056284
# Brier score                                  0.007994
# Predict time (s)                             0.007835
# Use .frame() to control the format of the output.
# Metric 'fail' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'accuracy' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'precision' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'recall' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'roc_auc' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'log_loss' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'brier_score' has failed: ValueError('No train data were provided when creating the report.')
# Metric 'fail' has failed: Exception('hey')
```

---

`CrossValidationReport`
```py
from sklearn.ensemble import RandomForestClassifier

import skore
import sklearn

X, y = sklearn.datasets.make_classification(n_samples=1000, return_X_y=True)

model = RandomForestClassifier()
report = skore.evaluate(model, X, y, splitter=3)

def fail(e,x,y):
    raise Exception("hey")

report.metrics.add(fail)

report.metrics.summarize()
#                        RandomForestClassifier
#                                          mean       std
# Metric           Label
# Fail                                      NaN       NaN
# Accuracy                             0.965993  0.006297
# Precision        0                   0.953693  0.019897
#                  1                   0.979622  0.008645
# Recall           0                   0.980004  0.009135
#                  1                   0.951964  0.021756
# ROC AUC                              0.989542  0.003003
# Log loss                             0.159800  0.054424
# Brier score                          0.030923  0.002292
# Fit time (s)                         0.137264  0.009510
# Predict time (s)                     0.004770  0.000060
# Use .frame() to control the format of the output.
# Metric 'fail' has failed: Exception('hey')
```

---

`ComparisonReport[CV]` where only one of the reports has the failing metric
```py
from sklearn.ensemble import RandomForestClassifier

import skore
import sklearn
import copy

X, y = sklearn.datasets.make_classification(n_samples=1000, return_X_y=True)

model = RandomForestClassifier()
report = skore.evaluate(model, X, y, splitter=3)

def fail(e,x,y):
    raise Exception("hey")

report_with_fail = copy.copy(report)

report_with_fail.metrics.add(fail)

skore.compare([report, report_with_fail]).metrics.summarize()
#                                            mean                                               std
# Estimator              RandomForestClassifier_1 RandomForestClassifier_2 RandomForestClassifier_1 RandomForestClassifier_2
# Metric           Label
# Fail                                        NaN                      NaN                      NaN                      NaN
# Accuracy                               0.926993                 0.926993                 0.009725                 0.009725
# Precision        0                     0.921393                 0.921393                 0.015023                 0.015023
#                  1                     0.933412                 0.933412                 0.019590                 0.019590
# Recall           0                     0.933975                 0.933975                 0.020880                 0.020880
#                  1                     0.919979                 0.919979                 0.017444                 0.017444
# ROC AUC                                0.968959                 0.968959                 0.010729                 0.010729
# Log loss                               0.249354                 0.249354                 0.075652                 0.075652
# Brier score                            0.059976                 0.059976                 0.006762                 0.006762
# Fit time (s)                           0.144223                 0.144223                 0.005398                 0.005398
# Predict time (s)                       0.004864                 0.004864                 0.000194                 0.000194
# Use .frame() to control the format of the output.
# Metric 'fail' has failed: Exception('hey')
```

</details>

### CODE REVIEW

File: skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py:1]: will conflict with #3094

File: skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py [hunk 1]
```diff
@@ -334,25 +342,34 @@ def _frame_cross_validation(
...
-
-        df = df.reset_index().pivot_table(
-            index=df.index.names,
-            columns="split" if aggregate is None else None,
-            values=estimator_name,
-            aggfunc="first" if aggregate is None else aggregate,
-            sort=False,
-        )
-
         if aggregate is None:
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py:347]: the main issue here was to ensure that the pivoting does not drop our failed metrics (which are NaN).

**Comment by jeromedockes** [comment]
thanks! from a quick look it LGTM but I'm not sure which we want to merge first this or 3094 / how to handle conflicts

**Comment by glemaitre** [comment]
Let's merge this PR and I will handle the conflict in #3094

**Review by jeromedockes — DISMISSED** [review]
thanks @auguste-probabl

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 2]
```diff
@@ -140,12 +144,28 @@ def summarize(
...
-                data_source=data_source,
-                **parsed_metric.kwargs,
-            )
+            try:
+                metric_rows = parsed_metric.rows(
+                    report=self._parent,
+                    data_source=data_source,
+                    **parsed_metric.kwargs,
+                )
+            except Exception as exception:
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:155]: I think we mentioned the possibility to issue a warning here? in addition to the message shown in the display, for users who don't look at the display but only get the frame()

Comment (jeromedockes) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py:155]: the warning could say how to remove the metrics to silence the warning

Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py:155]: done

**Review by jeromedockes — APPROVED** [review]
thank you @auguste-probabl !
